packages/expo-cli/src/commands/upload/submission-service/android/AndroidSubmitter.ts

```diff
@@ -0,0 +1,279 @@
+  androidPackageSource: AndroidPackageSource;
+  archiveSource: ArchiveSource;
+  serviceAccountSource: ServiceAccountSource;
+}
+
+interface ResolvedSourceOptions {
+  androidPackage: string;
+  archive: Archive;
+  serviceAccountPath: string;
+}
+
+class AndroidSubmitter {
+  constructor(private ctx: AndroidSubmissionContext, private options: AndroidSubmissionOptions) {}
+
+  async submitAsync(): Promise<void> {
+    const resolvedSourceOptions = await this.resolveSourceOptions();
+    if (this.ctx.mode === SubmissionMode.online) {
+      const submissionConfig = await AndroidOnlineSubmitter.formatSubmissionConfigAndPrintSummary(
+        this.options,
+        resolvedSourceOptions
+      );
+      const onlineSubmitter = new AndroidOnlineSubmitter(
+        submissionConfig,
+        this.ctx.commandOptions.verbose ?? false
+      );
+      await onlineSubmitter.submitAsync();
+    } else {
+      const submissionConfig = await AndroidOfflineSubmitter.formatSubmissionConfigAndPrintSummary(
+        this.options,
+        resolvedSourceOptions
+      );
+      const offlineSubmitter = new AndroidOfflineSubmitter(submissionConfig);
+      await offlineSubmitter.submitAsync();
+    }
+  }
+
+  private async resolveSourceOptions(): Promise<ResolvedSourceOptions> {
+    const androidPackage = await getAndroidPackageAsync(this.options.androidPackageSource);
+    const archive = await getArchiveAsync(this.ctx.mode, this.options.archiveSource);
+    const serviceAccountPath = await getServiceAccountAsync(this.options.serviceAccountSource);
+    return {
+      androidPackage,
+      archive,
+      serviceAccountPath,
+    };
+  }
+}
+
+interface AndroidOfflineSubmissionConfig
+  extends Pick<
+    AndroidSubmissionConfig,
+    'archiveType' | 'track' | 'releaseStatus' | 'androidPackage'
+  > {
+  archivePath: string;
+  serviceAccountPath: string;
+}
+
+class AndroidOfflineSubmitter {
+  static async formatSubmissionConfigAndPrintSummary(
+    options: AndroidSubmissionOptions,
+    { archive, androidPackage, serviceAccountPath }: ResolvedSourceOptions
+  ): Promise<AndroidOfflineSubmissionConfig> {
+    const submissionConfig = {
+      androidPackage,
+      archivePath: archive.location,
+      archiveType: archive.type,
+      serviceAccountPath,
+      ...pick(options, 'track', 'releaseStatus'),
+    };
+    printSummary({
+      ...omit(submissionConfig, 'serviceAccount'),
+      mode: SubmissionMode.offline,
+    });
+    return submissionConfig;
+  }
+
+  constructor(private submissionConfig: AndroidOfflineSubmissionConfig) {}
+
+  async submitAsync(): Promise<void> {
+    const {
+      archivePath,
+      archiveType,
+      androidPackage,
+      serviceAccountPath,
+      track,
+      releaseStatus,
+    } = this.submissionConfig;
+
+    // TODO: check if `fastlane supply` works on linux
+    const travelingFastlane = require(`@expo/traveling-fastlane-${os.platform()}`)();
```

Comment:
For windows we should use `@expo/traveling-fastlane-linux`

Reply:
Also some better error message could be useful here

Reply:
I changed that. Regarding the error message - there is already a meaningful error message - https://github.com/expo/expo-cli/pull/2069/files#diff-911f5cceaa622faf6d5d53c7b83e5d3dR128 Generally, this code will never be executed on a different platform than macOS (yet).